### PR TITLE
Escape userinfo in PostgreSQL DSN using net/url UserInfo

### LINKED ISSUE #61: PostgreSQL client fails to connect if password contains characters that need to be URL encoded
### What happened?

My (auto-generated) postgres password contained an '^' character. As a result, provider-sql was not able to connect to the database. The following error was present in the logs (password, host and dbname redacted):

Warning    {"object": {"kind":"Extension","name":"isvanilla-uuid-extension","uid":"9c74ceb9-5
f04-4b4c-b508-00a820cea033","apiVersion":"postgresql.sql.crossplane.io/v1alpha1","resourceVersion":"300047"}, "reason": "CannotObserveExternalResource", "message": "ca
nnot select extension: parse \"postgres://postgres:[password]@[host]:5432/[dbname]\": net/ur
l: invalid userinfo"}

Looking at the code that produces the DSN URL, I think there is a need to escape the string to ensure the username and password are safely encoded for use in a URL. 

```
func New(creds map[string][]byte, database string) xsql.DB {
	// TODO(negz): Support alternative connection secret formats?
	endpoint := string(creds[xpv1.ResourceCredentialsSecretEndpointKey])
	port := string(creds[xpv1.ResourceCredentialsSecretPortKey])
	return postgresDB{
		dsn: "postgres://" +
			string(creds[xpv1.ResourceCredentialsSecretUserKey]) + ":" +
			string(creds[xpv1.ResourceCredentialsSecretPasswordKey]) + "@" +
			endpoint + ":" +
			port + "/" +
			database,
		endpoint: endpoint,
		port:     port,
	}
}
```

### How can we reproduce it?

Attempt to connect to a Postgres database using a password containing a '^' character

### What environment did it happen in?

Crossplane version: 1.6.1
Running in EKS on AWS (K8s 1.21)

### PR DESCRIPTION
**Pull request by jdotw**
### Description of your changes

Fixes #61 

If a password contains a character that needs to be percent-encoded for use in the DSN, the connection to postgres will fail with a `net/url: invalid userinfo` error. 

This change first refactors the creation of the DSN string to allow unit testing, then uses net/url's UserInfo to encode the username:password portion of the DSN. This will apply percent-encoding appropriately. 

Included a unit test that checks for proper encoding of password with a '^' character which is valid in PostgreSQL, but causes provider-sql to fail to connect without this fix. 

I have:

- [X] Read and followed Crossplane's [contribution process].
- [X] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Added unit test to confirm string encoding

Ran 'make reviewable test'

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Review by Duologic — APPROVED** [review]
I'm running into the same issue, I have an `@` in the password. The net/url.UserPassword function is also what postgres_exporter uses. 

LGTM

**Comment by negz** [comment]
Closed and re-opened in the hopes of triggering CI.

File: pkg/clients/postgresql/postgresql.go [hunk 1]
```diff
@@ -30,18 +31,28 @@ func New(creds map[string][]byte, database string) xsql.DB {
...
-			port + "/" +
-			database,
+		dsn:      DSN(username, password, endpoint, port, database),
 		endpoint: endpoint,
 		port:     port,
 	}
 }
 
+// DSN returns the DSN URL
+func DSN(username string, password string, endpoint string, port string, database string) string {
```
Comment (negz) [inline pkg/clients/postgresql/postgresql.go:44]:
```suggestion
func DSN(username, password, endpoint, port, database string) string {
```

Just a nitpick. :)

**Review by negz — APPROVED** [review]

**Comment by negz** [comment]
Whoops - I may have approved too soon. Looks like a unit test is failing.

Also, would we need a similar change for MySQL?

**Comment by jdotw** [comment]
> Whoops - I may have approved too soon. Looks like a unit test is failing.
> 
> Also, would we need a similar change for MySQL?

Likely.. I've opened #70 to track this.

**Pull request merged**